File: NWMv4/scripts/gower_dist.R
# apply Gower distance measure to identify physically similar donors
gower_dist <- function(config, dtAttrAll, attr_scenario, dist_spatial) {
  
  #receiversAll <- dtAttrReceiver$id
  dtDonorAll <- data.table()
  
  # two rounds processing, first with attrs$main, and then with attrs/$base for catchments with no donors found in 1st round
  attrs1 <- list(main=config$attrs[[attr_scenario]], base=config$attrs$base)
  for (run1 in names(attrs1)) { 
    
    # check if donors are identified for all receivers
    recs0 <- subset(dtAttrAll, tag=="receiver")$id
    recs0 <- recs0[!recs0 %in% dtDonorAll$id]
    if (length(recs0)==0) next
    
    # get the attributes to use
    dtAttr0 <- dtAttrAll[,c(config$non_attr_cols,attrs1[[run1]]),with=F]
    
    # iteratively process all the receivers to handle data gaps
    recs1 <- NULL #recerivers that have already been processed in previous rounds
    kround <- 0
    while(sum(recs0 %in% recs1)!=length(recs0)) {
      
      kround <- kround + 1
      message(paste0("\n------------------------",run1, " attributes,  Round ",kround,"--------------------"))
      
      # perform pca to obtain the scores and weights (as well as the attribute table used for PCA)
      pca1 <- apply_pca(recs0, recs1,dtAttr0, attrs1[[run1]])
      
      # donors and receivers for this round
      donorsAll1 <- subset(pca1$dtAttr,tag=="donor")$id
      receiversAll1 <- subset(pca1$dtAttr,tag=="receiver")$id
      
      # compute Gower's distance between donors and receivers only, i.e., avoid calculating distance
      # between donors and donors, receivers and receivers (faster)
      system.time({      
        nd1 <- length(donorsAll1)
        nr1 <- length(receiversAll1)
        rng1 <- sapply(colnames(pca1$scores), function(x) diff(range(pca1$scores[,x])))
        rng2 <- matrix(rep(rng1,nr1), nrow=nr1, byrow=TRUE)
        wgt2 <- matrix(rep(pca1$weights,nr1),nrow=nr1, byrow=TRUE)
        distAttr0 <- sapply(1:nd1, function(x) rowSums(abs(pca1$scores[(nd1+1):nrow(pca1$scores),]-
                                                             matrix(rep(pca1$scores[x,],nr1),nrow=nr1,byrow=TRUE))/rng2*wgt2))
      })     
      
      # determine which receivers to be processed for the current round
      # process only those not-yet processed receivers
      recs <- recs0[(recs0 %in% receiversAll1) & (!recs0 %in% recs1)]
      recs1 <- c(recs1, recs)
      message(paste0(length(recs)," receivers to be processed this round"))      
      
      # loop through all the receivers to be processed
      for (rec1 in recs) {
        
        # find donors within the defined buffer iteratively so that donors with attr distance
        # below the predefined value can be found
        buffer <- config$pars$gower$minSpaDist - 100
        while(buffer < (config$pars$general$maxSpaDist -100)) {
          
          buffer <- buffer + 100
          
          # if there exists donor catchment within 1km, select that catchment as donor
          donors1 <- names(which(dist_spatial[rec1,donorsAll1]<=1))
          if (length(donors1)>0) {
            donors1 <- donors1[which.min(dist_spatial[rec1,donors1])]
          } else {
            # narrow down to donors within the buffer
            donors1 <- names(which(dist_spatial[rec1,donorsAll1]<= buffer))
          }
          
          # potential donors with dist <= config$pars$gower$maxAttrDist
          dist1 <- distAttr0[match(rec1,receiversAll1),]
          dist1 <- dist1[match(donors1,donorsAll1)]
          ix1 <- which(dist1 <= config$pars$gower$maxAttrDist)
          if (length(ix1)==0) next
          
          # if a suitable donor is found, break the loop and stop searching
          if (min(dist1[ix1]) <= config$pars$gower$minAttrDist | length(donors1)==length(donorsAll1)) break
        }
        
        donor1 <- NULL
        if (length(donors1)>0) {
          
          # donors that satisfy the config$pars$gower$maxAttrDist threshold
          dist1 <- dist1[ix1]
          donor1 <- donors1[ix1]
          
          # apply additional donor constraints
          list1 <- apply_donor_constraints(rec1, donor1, dist1, config$pars, dtAttrAll)
          dist1 <- list1$dist; donor1 <- list1$donor
          
          if (length(donor1)>0) {
            
            # choose the donors with the smallest attribute distance
            ix2 <- order(dist1)[1:min(length(dist1),config$pars$gower$nDonorMax)]
            distAttr1 <- dist1[ix2]
            donor1 <- donor1[ix2]
            donor1 <- donor1[!is.na(donor1)]
            
            # donor-receiver spatial distance
            distSpatial1 <- dist_spatial[rec1,donor1]
            
            # choose the donor with the smallest spatial distance (best donor)
            if (length(donor1)>0) {
              ix3 <- which.min(distSpatial1)
              donor_best1 <- donor1[ix3]
              distAttr_best1 <- distAttr1[ix3]
              distSpatial_best1 <- distSpatial1[ix3]
            }}
        }
        
        # if no donors found, get the spatially closest donor in the same HSG
        tag1 <- run1
        if (length(donor1)==0 & run1=="base") {
          
          # get all donors and their spatial distance to the receiver
          donor0 <- subset(dtAttrAll, tag=="donor")$id
          dist0 <- dist_spatial[rec1,]
          
          # apply additional donor constraints
          list1 <- apply_donor_constraints(rec1, donor0, dist0, config$pars, dtAttrAll)
          dist0 <- list1$dist; donor0 <- list1$donor
          
          # choose the donor that is spatially closest
          donor1 <- donor0[which.min(dist0)]
          distSpatial1 <- min(dist0)
          distAttr1 <- distAttr0[match(rec1,receiversAll1),match(donor1,donorsAll1)]
          tag1 <- "proximity"
          
          # best donor (same as donor1)
          donor_best1 <- donor_best1
          distSpatial_best1 <- distSpatial1
          distAttr_best1 <- distAttr1
        }
        
        if (length(donor1)>0) {
          df1 <- data.frame(id=rec1, tag=tag1,
                            donor=donor_best1,distAttr=round(distAttr_best1,3),distSpatial=distSpatial_best1,
                            donors=paste(donor1,collapse=","), 
                            distAttrs=paste(round(distAttr1,3),collapse=","),
                            distSpatials=paste(distSpatial1,collapse=","))
          dtDonorAll <- rbind(dtDonorAll, as.data.table(df1))
        } 
      } # loop receiver
    } # loop round
    
    #kround <- kround + 1
    
  } #loop attrs main + base
  
  # print summaries
  message("\nNumber of receivers in each category:")
  print(table(dtDonorAll$tag))
  message("\nSummary of Gower's distance:")
  message("==== main ====")
  print(summary(subset(dtDonorAll,tag=="main")$distAttr))
  message("==== base ====")
  print(summary(subset(dtDonorAll,tag=="base")$distAttr))
  message("==== proximity ====")
  print(summary(subset(dtDonorAll,tag=="proximity")$distAttr))
  
  return(dtDonorAll)
  
} # function gower

File: NWMv4/algorithm/gower_dist.py
def func(config, dtAttrAll,scenario, dist_spatial):
    print("calling function gower_dist ...")

    import pandas as pd
    import numpy as np
    import apply_pca
    import apply_donor_constraints

    dtDonorAll = pd.DataFrame()

    # two rounds of processing, first with attrs defined by the selected scenario (i.e., 'main'), 
    # and then with 'base' attrs for catchments with no donors found in 1st round
    attrs1 = {'main': config['attrs'][scenario],
            'base': config['attrs']['base']}
    for run1 in attrs1: # attr round
        # check if donors are identified for all receivers
        recs0 = dtAttrAll.query("tag=='receiver'")['id'].values
        if dtDonorAll.shape[1] > 0:
            recs0 = [value for value in recs0 if value not in dtDonorAll["id"].values]
        if len(recs0)==0:
            continue
    
        # reduce the attribute table to attributes for the current round
        dtAttr0 = dtAttrAll[config['non_attr_cols']+attrs1[run1]]

        # iteratively process all the receivers to handle data gaps so that 
        # receivers with the same missing attributes are processed in the same round
        recs1 = list() #receivers that have already been processed in previous krounds
        kround = 0
        while len([x for x in recs0 if x in recs1]) != len(recs0):
            kround = kround + 1
            print("\n------------------------" + run1 + " attributes,  Round " + str(kround) + "--------------------")   
            #pdb.set_trace()
            
            # apply principal component analysis
            myscores, weights, dtAttr = apply_pca.func(recs0, recs1, dtAttr0, attrs1[run1])
            
            # donors and receivers for this round
            donorsAll1 = dtAttr.query("tag=='donor'")['id'].tolist()
            receiversAll1 = dtAttr.query("tag=='receiver'")['id'].tolist()

            # compute Gower's distance between donors and receivers only, i.e., avoid calculating distance
            # between donors and donors, receivers and receivers (faster)     
            nd1 = len(donorsAll1)
            nr1 = len(receiversAll1)
            rng1 = myscores.max() - myscores.min()
            rng2 = np.repeat(np.matrix(rng1),nr1,axis=0)
            wgt2 = np.repeat(np.matrix(weights),nr1,axis=0)
            distAttr0 = pd.DataFrame()
            scores_receiver = myscores.iloc[nd1:]
            for r1 in range(nd1):
                scores_donor = np.repeat(np.matrix(myscores.iloc[r1]),nr1,axis=0)
                df1 = ((scores_donor - scores_receiver).abs()/rng2*wgt2).sum(axis=1)
                distAttr0 = pd.concat((distAttr0,df1),axis=1)
            distAttr0.columns = donorsAll1
            distAttr0.index = receiversAll1

            # determine which receivers to be processed for the current round
            # process only those not-yet processed receivers
            recs = [r1 for r1 in recs0 if r1 in receiversAll1 and r1 not in recs1]
            recs1 = recs1 + recs
            print(str(len(recs)) + " receivers to be processed this round")     

            # loop through all the receivers to be processed
            for rec1 in recs:

                # find donors within the defined buffer iteratively so that closest donors 
                # with attr distance below the predefined value can be found
                buffer = config['pars']['gower']['minSpaDist'] - 100
                while buffer < config['pars']['general']['maxSpaDist'] -100 :

                    buffer = buffer + 100
                    # print("buffer = " + str(buffer))
                    
                    #if there exists donor catchment within a short distance,               
                    s1 = dist_spatial.loc[rec1]
                    donors1 = s1.loc[s1<=config['pars']['gower']['zeroSpaDist']].index.tolist()                    
                    if len(donors1)>0 :
                        # select that catchment as donor;
                        donors1 = [s1[donors1].idxmin()]
                    else:
                        # otherwise, narrow down to donors within the buffer
                        donors1 = s1.loc[s1<=buffer].index.tolist()
                                
                    # potential donors with dist <= maxAttrDist
                    dist1 = distAttr0.loc[rec1, donors1]
                    ix1 = [x for x in dist1.index if dist1[x] <= config['pars']['gower']['maxAttrDist']]
                    if (len(ix1)==0): # if not, continue to the next round with a larger neighbourhood
                        continue

                    # if a suitable donor is found (or all donors have been assessed), break the loop and stop searching
                    if dist1[ix1].min() <= config['pars']['gower']['minAttrDist'] or len(donors1)==len(donorsAll1): 
                        break

                # if donors are identified
                if len(donors1)>0:   
                                        
                    # narrow down to those that satisfy the maxAttrDist threshold
                    dist1 = dist1.loc[ix1]
                    donor1 = dist1.index.tolist()
                
                    # apply additional donor constraints
                    donor1, dist1 = apply_donor_constraints.func(rec1, donor1, dist1, config['pars'], dtAttrAll)
                
                    # if donors remain
                    if len(donor1)>0:          
                        # choose those with the smallest attribute distance
                        ix2 = np.argsort(dist1)[range(min(len(dist1),config['pars']['gower']['nDonorMax']))] 
                        distAttr1 = dist1[ix2]
                        donor1 = donor1[ix2]            
                
                        # get donor-receiver spatial distance
                        distSpatial1 = np.array(dist_spatial.loc[rec1,donor1])
                
                        # choose the donor with the smallest spatial distance (best donor)
                        if len(donor1)>0:
                            ix3 = np.where(distSpatial1 == distSpatial1.min())[0][0] #"[0][0]" gets index of 1st minimum if multiple minima exist                                               
                            donor_best1 = donor1[ix3].item()
                            distAttr_best1 = distAttr1[ix3].item()
                            distSpatial_best1 = distSpatial1[ix3].item()
                
                # if no donors found (after both 'main' and 'base' attribute rounds), 
                # get the spatially closest donor with some constraints
                tag1 = run1
                if len(donor1)==0 and run1=="base":                       
                    # get all donors and their spatial distance to the receiver
                    donor0 = dtAttrAll.query("tag=='donor'")['id']
                    dist0 = dist_spatial.loc[rec1,]
                
                    # apply additional donor constraints
                    donor0, dist0 = apply_donor_constraints.func(rec1, donor0, dist0, config['pars'], dtAttrAll)
                
                    # choose the donor that is spatially closest
                    donor_best1 = donor1 = donor0[np.where(dist0==min(dist0))]
                    distSpatial_best1 = distAttr1 = min(dist0)
                    distAttr_best1 = distSpatial1 = distAttr0.loc[rec1,donor1]
                    tag1 <- "proximity"
                
                # add the donor/receiver pair to the pairing table    
                if len(donor1)>0:
                    pair1 = {'id': rec1, 'tag': tag1, 'donor': donor_best1,
                        'distAttr': np.round(distAttr_best1,3),
                        'distSpatial': distSpatial_best1,
                        'donors': ','.join(donor1), 
                        'distAttrs': ','.join(map(str,pd.Series(np.round(distAttr1,3)))),
                        'distSpatials': ','.join(map(str,pd.Series(distSpatial1)))}
                    dtDonorAll = pd.concat((dtDonorAll, pd.DataFrame(pair1,index=[0])),axis=0)
    
    return dtDonorAll
